# Mapping reads from samples to the concatenated human-target nuclear reference genomes
 - The target species portion of the nuclear genome has had mitochondrial-like regions masked (see folder `1-DataPrep\1.3-MaskMitoRegions`).
 - Only one sample per reference genome is included here as a representative of the rest, as the only thing that changes is the SampleID and the read file names.
 - Script `2.2.1-compMap_nc_array.slurm` is only the job submission script that submits all the actual scripts, which are here represented by the `compMap_nc_*.sh` scripts, which in turn use the relevant `*.yaml` file.

## Masking microbial-like regions in BAM files
 - The script `2.2.2-subset_BAMs_NonMicrobeRegions.slurm` uses the BED files to subset/mask the BAM files of all samples by removing reads overlapping with the regions in the BED files. 
 - The microbial-like regions were extracted from the large BED file from Oskolkov et al. (2025) for the relevant assembly.
 - For the common eland and Cape grysbok, the microbial-like BED files were generated by us (see folder `1-DataPrep\1.4-DetectMicroRegions`).
 - This produces the `*noMicro.bam` files, which are the final BAM files used for endogenous DNA estimation and DamageProfiler analysis.